REACT_STYLEGUIDE.MD

```diff
@@ -0,0 +1,203 @@
+
+Good:
+```
+<hello-component fname="person.fname" lname="person.lname" watch-depth="reference"></hello-component>
+```
+Bad:
+```
+<react-component name="HelloComponent" props="person" watch-depth="reference" />
+```
+
+### Prefix ui_framework elements with kui, but not their file names.
+Good:
+```
+button.js:
+export function KuiButton(props) {
+  return <button className="kuiButton" { ...props } />
+};
+```
+Bad:
+```
+button.js:
+export function Button(props) {
+  return <button className="kuiButton" { ...props } />
+};
+```
+The filenames leave it off because snake casing already increases file name length.
+
+### Action function names and prop function names
+
+Name action functions in the form of a strong verb and passed properties in the form of on<Subject><Change>. E.g:
+```
+<sort-button onClick={ action.sort }/>
+<pagerButton onPageNext={ action.turnToNextPage } />
+```
+
+### Avoid creating a function and passing that as a property, in render functions.
+Best (relies on [stage 2 proposal](https://github.com/tc39/proposal-class-public-fields)):
+```
+export class ClickCounter extends React.Component {
+  state = { clickCount: 0 };
+
+  // This syntax ensures `this` is bound within handleClick
+  onClick = () => {
+    this.setState(prevState => { clickCount: prevState.clickCount + 1 });
+  }
+
+  render() {
+    return <button className="kuiButton" onClick={ this.onClick } />
+  }
+}
+```
+Good:
+```
+export class ClickCounter extends React.Component {
+  constructor() {
+    this.state = { clickCount: 0 };
+    this.onClick = this.onClick.bind(this);
+  }
+
+  onClick() {
+    this.setState(prevState => { clickCount: prevState.clickCount + 1 });
+  }
+
+  render() {
+    return <button className="kuiButton" onClick={ this.onClick } />
+  }
+}
+```
+
+Bad:
+```
+export class ClickCounter extends React.Component {
+  state = { clickCount: 0 };
+
+  onClick() {
+    this.setState(prevState => { clickCount: prevState.clickCount + 1 });
+  }
+
+  render() {
+    return <button className="kuiButton" onClick={ () => this.onClick() } />
+  }
+}
+```
+Background: https://facebook.github.io/react/docs/handling-events.html
+There is also an eslint rule we should be able to turn on for this.
+
+### Never mutate state directly
+Good:
+```
+this.setState(prevState => { clickCount: prevState.clickCount + 1 });
+```
+Bad:
+```
+this.state.clickCount += 1;
+```
+
+### Prefer primitives over objects when storing in state.
+Good:
+```
+this.setState({
+  currentPage: 0,
+  selectedIds: []
+});
+```
+
+Discouraged:
+```
+this.setState({
+  pager: new Pager(),
+  selectedIds: new SelectedIds()
+});
+```
+
+### Favor spread operators
+Good:
+```
+render() {
+  return <button className="kuiButton" { ...this.props } />
+}
+```
+
+Bad:
+```
+render() {
+  return <button className="kuiButton" onClick={ this.props.onClick } isDisable={ this.props.isDisabled } />
```

Comment:
Hm, do we really want to call this "Bad"? I definitely don't agree on that. I think this is good and clear. Imo the "Good" solution above is e.g. bad when you pass props that aren't handled by the component. (When we do that with html elements (like `button`) React will tell you if you pass it props it doesn't handle, but other (aka all non-React-core) components don't.)

I think this works well in some contexts (e.g. the tons of tiny simple wrapper components in the UI framework), but not in others.